README.md
# floating_action_wheel

<h1>
Floating Action Wheel
</h1>

<h2>
A new Floating Action button widget for Flutter with a multi-functional, immersive buttons wheel 
</h2>

## Getting Started


## 💻 Installation
Add the following dependency to your  `pubspec.yaml`

```yaml
dependencies:
  floating_action_wheel: <latest version>
```


## Usage:
You have two options to use the FloatingActionWheel, you can use it as a normal FAB in Scaffold(floatingActionButton) or you can 
put it in other compatible widgets like a Container or a Stack

Simply import the main classes and use the main FloatingActionWheel() constructor with its parameters, the mandatory parameter is the
buttons list where you pass a list of WheelButton each has its own properties, like its text, icon, image and colors, and its OnPressed callback. Look up the internal documentation for the WheelButton class and its parameters




```dart
import 'package:floating_action_wheel/floating_action_wheel.dart';
import 'package:floating_action_wheel/wheel_button.dart';
```

### - Implementation
```dart
FloatingActionWheel(
      buttons: [
        WheelButton(onPressed: () {
          setState(() {
            demoText = "button #1 pressed";
            backgroundColor= Colors.orangeAccent;
          });
        },
            text: 'button 1',
            backgroundColor: Colors.orange),
        WheelButton(onPressed: () {
          setState(() {
            demoText = "button #2 pressed";
            backgroundColor= Colors.greenAccent;
          });
        },
            icon: Icons.ac_unit,
            backgroundColor: Colors.green),
            
        WheelButton(onPressed: () {
          setState(() {
            demoText = "button #3 pressed";
            backgroundColor= Colors.cyanAccent;
          });
        },
            image: Image.asset('assets/your_image.png'),
            backgroundColor: Colors.cyan),
            
        WheelButton(
            onPressed: () {
          setState(() {
            demoText = "button #4 pressed";
            backgroundColor= Colors.pinkAccent;
          });
        },
            icon: Icons.home,
            backgroundColor: Colors.pink),

      ],
      angleOffset: 90,
      visiblePart: 0.5,
      animationType: WheelAnimationType.around,
      wheelSize: WheelSize.wheel_medium_120,
      fabForegroundColor: Colors.white,
      fabBackgroundColor: Colors.blue,
      separated: false,
    );
) 
```

### - Demonstration example
<img src=
"https://user-images.githubusercontent.com/75282663/139159039-4d6e705c-6fcf-48cc-ae4f-6d269a8f5c2c.jpg"
height = "320" width = "180" />
<img src=
"https://user-images.githubusercontent.com/75282663/139159070-256fa6ee-6110-4351-b596-63fd08a31eca.jpg"
height = "320" width = "180" />
<img src=
"https://user-images.githubusercontent.com/75282663/139159080-89dce8fe-4873-4549-bbc6-34dcd926a9f9.jpg"
height = "320" width = "180" />
<img src=
"https://user-images.githubusercontent.com/75282663/139159089-20e5ac97-4fa5-40f1-bd98-adf9a6601370.jpg"
height = "320" width = "180" />
<img src=
"https://user-images.githubusercontent.com/75282663/139159099-6bd7d09c-3b33-4744-a1ed-4f7e42458bc7.jpg"
height = "320" width = "180" />
<img src=
"https://user-images.githubusercontent.com/75282663/139159106-7be4851f-d03f-41d9-a1cc-5949dfd18ce4.jpg"
height = "320">




CHANGELOG.md
## v[1.0.1] - release - sound null safety

Initial release: v1.0.1
Features/changes from previous versions:
1- Upgraded to null safety, now compatible with all null safety Flutter projects
2- Added new feature to draw images on wheel buttons instead of text or icon
2- Added new feature to instantiate the FAW widget with wheel already shown
3- Fixed minor bugs in Demo app



## v[0.9.1] - Beta for open testing

Initial beta release: v0.9.1 open BETA
features/changes from previous versions:
1- the FAB and the ImmersiveWheel widgets now using CompositeTransformTarget/Follower widgets to attach them to each other, previous version
   was using an Overlay widget and was incompatible with the Scaffold's FAB slot 
2- added parameter to separate each button
3- added text/icon parameters to wheel buttons as well as giving each its size explicitly if they dont fit well



test/floatingactionwheel_test.dart
import 'dart:math';


import 'package:floating_action_wheel/src/wheel_slice.dart';
import 'package:floating_action_wheel/floating_action_wheel.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';



void main() {

  testWidgets('Counter increments smoke test', (WidgetTester tester) async {

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });

  int x = 10;
  var y = 'kkk';

  List<WheelButton> wheelbtns = [
    WheelButton(
        onPressed: () => x += 10,
        text: 'text1',
        textSize: 12,
        backgroundColor: Colors.red),
    WheelButton(
        onPressed: () => x -= 10,
        text: 'text2',
        icon: Icons.home,
        iconSize: 13,
        backgroundColor: Colors.blue),
    WheelButton(
        onPressed: () => y += '10',
        text: 'text3',
        textSize: 11,
        icon: Icons.ac_unit,
        iconSize: 15,
        backgroundColor: Colors.green),
    WheelButton(
        onPressed: () => y += '20',
        text: 'text4',
        backgroundColor: Colors.yellow,
        foregroundColor: Colors.yellowAccent),
  ];

  var mainWheel = FloatingActionWheel(key: Key('ok'),
    buttons: wheelbtns,
    wheelSize: WheelSize.wheel_large_150,
    animationType: WheelAnimationType.none,
    visiblePart: 1,
    angleOffset: 90,
    separated: false,
    fabBackgroundColor: Colors.brown,
    fabForegroundColor: Colors.black,
    fabHeroTag: 'tag',
  );


  group('main widget test', () {
    testWidgets('main base test', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          floatingActionButton: mainWheel,
          floatingActionButtonLocation: FloatingActionButtonLocation
              .centerDocked,
          body: Center(
            child: Text('wheel test'),
          ),
        ),
      )
      );

      expect(find.byType(FloatingActionButton),findsOneWidget);
    });

    testWidgets('FAB test', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          floatingActionButton: mainWheel,
          floatingActionButtonLocation: FloatingActionButtonLocation
              .centerDocked,
          body: Center(
            child: Text('wheel test'),
          ),
        ),
      )
      );

      expect(find.byType(FloatingActionButton),findsOneWidget);
//      expect(find.byElementType(FloatingActionButton),findsOneWidget);
    });


    testWidgets('FAB test2', (WidgetTester tester) async {
      List<int> x=[4,2,6,7,4,3];
      List<int> b= new List.empty();
//      = x.map(( int i) => i+ (i-1) +(i+1)).toList();

      String n='450';
      String nb='44';
      String nm='377';

      int k= int.tryParse(n)!;
      int j= Random(1).nextInt(n.length);
      String kk=n.replaceRange(j, j+1, '');

      if(kk.length >1 && !kk.startsWith("0")) {
        if (int.tryParse(nm) == int.tryParse(nb)! + int.tryParse(kk)!)
          print("YES");
        else
          print('NO');
      }
      else
        print('NO');

//      for(int i=0;i<x.length;i++) {
//        b[i] = x[i] + (i <= 0 ? 0 : x[i-1]) + (i >= x.length -1 ? 0 : x[i+1]);
//      }
//      x.forEach((element) {int x = element + (x-)})
      print(b);

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          floatingActionButton: mainWheel,
          floatingActionButtonLocation: FloatingActionButtonLocation
              .centerDocked,
          body: Center(
            child: Text('wheel test'),
          ),
        ),
      )
      );
      await tester.tap(find.byType(FloatingActionButton));

//      expect(find.byKey(Key('btn0')),findsOneWidget);
      expect(find.byType(WheelSlice),findsOneWidget);
//      await tester.tap(find.byKey(Key('btn2')));

//      print(x);


      expect(find.byType(FloatingActionButton),findsOneWidget);
//      expect(find.byElementType(FloatingActionButton),findsOneWidget);
    });



  });
}


example/lib/main.dart

import 'package:flutter/material.dart';

import 'demo_route.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Floating Action Wheel Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home:DemoRoute()
      // DismissKeyboard(
      //     child:
    // ),
    );
  }
}



//TODO not working
class DismissKeyboard extends StatelessWidget {
  final Widget child;
  const DismissKeyboard({Key? key, required this.child}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        FocusScopeNode currentFocus = FocusScope.of(context);
        if (
            currentFocus.focusedChild != null) {
          print('unfocused');
          FocusManager.instance.primaryFocus?.unfocus();
        }
      },
      child: child,
    );
  }
}

lib/src/wheel_slice.dart

import 'dart:async';
import 'dart:math' as math;
import 'dart:ui' as ui;
import 'package:flutter/material.dart' ;
import 'package:flutter/widgets.dart';



/// the Widget that holds an actual button and Arc Painter
class WheelSlice extends StatefulWidget {

  final String? text;
  final IconData? icon;
  final Color backgroundColor;
  final Function onPressed;
  final Color foregroundColor;


  final double stroke;
  final double startAngle;
  final double sweepAngle;
  final double? textSize;
  final double? iconSize;
  final double activePart;
  final ImageInfo? image;
  final int index;



  WheelSlice(this.startAngle, this.sweepAngle, this.index,
      {Key? key, this.text, this.icon, this.backgroundColor = Colors.blue,
        this.foregroundColor = Colors.white,
        this.textSize, this.iconSize,this.image,
        required this.onPressed,this.stroke=120,this.activePart=1}):
        assert(image == null || (icon == null && text == null));



  @override
  State<StatefulWidget> createState() => new WheelSliceState();

}

class WheelSliceState extends State<WheelSlice> {

  bool isPressed = false;

  late  double fontSize;
  late  double iconSize;
  late  Color _currentColor;
  late Color textIconColor;
  late Color _highlightColor;
  var gesture;

  @override
  void initState() {
    super.initState();
    _currentColor = widget.backgroundColor;

    _highlightColor = Color.alphaBlend(Color(0x70FFFFFF), _currentColor);

    textIconColor = widget.foregroundColor;
//    textIconColor = Color.alphaBlend(Color(0xAAFFFFFF), widget.backgroundColor);

    /**init for default icon size if no text provided **/
    fontSize = widget.stroke / 5;

    if (widget.text != null) {
      if (widget.textSize != null)
        fontSize = widget.textSize!;
      else {
        if (widget.stroke <= 90)
          fontSize = 20.0 - widget.text!.length * 0.8;
        else if (widget.stroke <= 120)
          fontSize = 26.0
              - math.pow(widget.text!.length, 0.8);
        else
          fontSize = 32.0 - widget.text!.length * 0.6;

        fontSize = fontSize * (math.pow(2,widget.activePart)/2);
      }
    }

    iconSize = widget.iconSize ?? fontSize;



    /**special gesture detection to handle proper highlight effect
        note: implementing onLongPress callbacks causes the quick tap highlight effect not to work **/
    gesture = GestureDetector(
      onTap: () {
        widget.onPressed();
        setState(() {
          isPressed =true;
        });
        //needed to force highlight effects on quick taps
        Timer(Duration(microseconds: 500), () =>
            setState(() {
              isPressed = false;
            }));
      },
      onTapDown: (_) {
        setState(() {
          isPressed = true;
        });
      },
      onTapCancel: () {
        setState(() {
          isPressed = false;
        });
      },
    );


  }


  @override
  Widget build(BuildContext context) {




    return
      ClipPath(
        clipper: ArcClip(widget.startAngle, widget.sweepAngle, widget.index),
        child: CustomPaint(
            painter: ArcPainter(widget.startAngle, widget.sweepAngle,
                isPressed ? _highlightColor : _currentColor,
                widget.index, widget.stroke,fontSize,iconSize,
                text: widget.text,
                icon: widget.icon,
                image:widget.icon ==null?widget.image:null,
                foregroundColor: textIconColor),
            child: gesture
        ),
      );
  }
}













/// the Custom painter to draw an Arc on the wheel
class ArcPainter extends CustomPainter {


  final double stroke;
  final double fontSize;
  final double iconSize;
  final double _startingAngle;
  final double _offsetAngle;
  Color paintColor;
  int index;
  final Color foregroundColor;
  final String? text;
  final IconData? icon;
  final ImageInfo? image;
  final double h=30,w=30;

//  final int wheelSize;

  int get getIndex => index;


  ArcPainter(this._startingAngle, this._offsetAngle, this.paintColor, this.index,
      this.stroke, this.fontSize,this.iconSize,
      {this.text, this.icon,this.image, this.foregroundColor = Colors.white});

  /// painting arc button on screen,
  /// creating new paint object based on given color and draw arc with zero based rects **/
  @override
  void paint(Canvas canvas, Size size) {
    final side = size.width < size.height ? size.width : size.height;
    final paint = Paint();

    paint.color = paintColor;
    paint.style = PaintingStyle.stroke;
    paint.strokeWidth = stroke;


    Rect _rect = new Rect.fromLTWH(0.0, 0.0, side, size.height);

    canvas.drawArc(
        _rect,
        _startingAngle,
        _offsetAngle,
        false,
        paint
    );


    /** offset from origin **/
    double offset = stroke == 180.0 ? 2 : 4;

    /** determine arc center points for drawing text/icon**/
    double radius = side / 3 + offset;

    final x = (radius) * math.cos(_startingAngle + (_offsetAngle / 2));
    final y = (radius) * math.sin(_startingAngle + (_offsetAngle / 2));

    if (image != null) {
      paintImage(canvas: canvas, rect:
      Rect.fromLTWH(
          (x + side / 2) - (h / 2),
          (y + side / 2) - (w / 2),
          h,
          w),
          image: image!.image,
          // fit: BoxFit.fill,
          scale: (stroke / 100) + 1);
    }
    else {

      /* draw text and/or icon on (x,y) */
      if (text != null && icon != null) {
        _drawText(canvas, text!, x, y - fontSize / 14, side, foregroundColor);
        _drawIcon(canvas, icon!, x, y + iconSize / 14, side, foregroundColor);
      }
      else if (text != null)
        _drawText(canvas, text!, x, y, side, foregroundColor);
      else if (icon != null)
        _drawIcon(canvas, icon!, x, y, side, foregroundColor);
    }
  }





  void _drawIcon(Canvas canvas, IconData icon, double x, double y, double side,
      Color color) {
    TextPainter tp = TextPainter(textDirection: TextDirection.ltr);
    tp.text = TextSpan(
      text: String.fromCharCode(icon.codePoint),
      style: TextStyle(
        color: color,
        fontSize: iconSize ,
        fontFamily: icon.fontFamily,
        package: icon.fontPackage, // This line is mandatory for external icon packs
      ),
    );
    tp.layout();
    tp.paint(canvas,
      new Offset(
        (side / 2 + x) - (tp.width / 2),
        text != null ? (side / 2 + y) - (tp.height / 2) + iconSize /2
            :
        (side / 2 + y) - (tp.height / 2),
      ),);
  }



  void _drawText(Canvas canvas, String name, double x, double y, double side,
      Color color) {

    TextStyle style = TextStyle(color: color, fontSize: fontSize);

    TextSpan span = TextSpan(style: style, text: name);

    TextPainter tp = TextPainter(
      text: span,
      textAlign: TextAlign.center,
      textDirection: TextDirection.ltr,
      textWidthBasis: TextWidthBasis.longestLine,
      maxLines: 2,
    );
    tp.layout(maxWidth: side);


    tp.paint(
      canvas,
      new Offset(
        (side / 2 + x) - (tp.width / 2),
        icon != null ? (side / 2 + y) - (tp.height / 2) - fontSize / 3
            :
        (side / 2 + y) - (tp.height / 2),
      ));

  }


  @override
  bool shouldRepaint(covariant ArcPainter oldDelegate) =>
      oldDelegate.paintColor != paintColor;
}




///
/// the CustomClipper which draws the clip to define the wheel button's edges
/// for touch pointers detection
class ArcClip extends CustomClipper<Path> {

  Path path = Path();


  final double _startAngle;
  final double _offsetAngle;
  int index;

  int get getIndex => index;

  Path get getpath => path;


  ArcClip(this._startAngle, this._offsetAngle, this.index,);


  @override
  Path getClip(Size size) {
    path = new Path()
      ..moveTo(size.width / 2, size.height / 2);
    path.lineTo(size.height, size.width);
    path.addArc(
        new Rect.fromLTWH(0.0, 0.0, size.height, size.width),// change H,W for perspective
        _startAngle,
        _offsetAngle);
    path.lineTo(size.height / 2, size.width / 2);
    path.close();
    return path;
  }


  @override
  bool shouldReclip(covariant ArcClip oldClipper) {
    return oldClipper.getpath != this.getpath;
  }

}


lib/src/wheel_button.dart



import 'package:flutter/material.dart';



/// Model class for a wheel button, serves only as an encapsulation for each of the wheel
/// buttons' properties
/// Parameters:
///   String [text] (Nullable) the text to display on the button
///   IconData [icon] (Nullable) the icon to show on the button
///   Color [backgroundColor] (Nullable) Button's color
///   Function [onPressed] (Non-Null) the functionality for the button's onPressed callback
///   Color [foregroundColor] (Nullable) the color for [text]/[icon]
class WheelButton {


  WheelButton({Key? key, this.text, this.icon,
    this.textSize, this.iconSize,this.image,
    this.backgroundColor=Colors.blue, this.foregroundColor=Colors.white, required this.onPressed}):
        assert(textSize != null ? text != null : textSize==null, 'Text size given without text'),
        assert(image == null || (icon == null && text == null),'Please either pass an image OR a text/icon'),
        assert(iconSize != null ? icon != null : iconSize==null, 'Text size given without text');


  ///   String? [text] (Nullable) the text to display on the button
  final String? text;


  ///   Image? [image] (Nullable) pass an [Image] object to draw the image on the button
  ///   Should only use image parameter without text or icon parameters
  final Image? image;


  ///   IconData? [icon] (Nullable) the icon to show on the button
  final IconData? icon;


  ///   Color [backgroundColor] (Nullable) Button's color
  ///
  ///   Defaults to [Colors.blue]
  final Color backgroundColor;


  ///   Color [foregroundColor] (Nullable) the color for [text]/[icon]
  ///
  ///   Defaults to [Colors.white]
  final Color foregroundColor;


  ///   Function [onPressed] (required) the functionality for the button's onPressed callback
  final Function onPressed;


  ///   Double [textSize] (Nullable) text font size for the [TextPainter]
  ///
  ///   Default size calculated based on wheel size
  final double? textSize;

  ///   Double [iconSize] (Nullable) icon size for the [TextPainter]
  ///
  ///   Default size calculated based on wheel size
  final double? iconSize;


  /* Member to hold the ImageInfo object to pass to the underlying CustomPaint in WheelSlice */
  ImageInfo? _img;

  set imageInfo(ImageInfo? value) {
    _img = value;
  }

  ImageInfo? get img => _img;



}





example/lib/demo_route.dart

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:floating_action_wheel/floating_action_wheel.dart';


class DemoRoute extends StatefulWidget{

  DemoRoute({Key? key}):super(key: key);


  @override
  State<StatefulWidget> createState() {
    return new DemoRouteState();
  }


}

class DemoRouteState extends State<DemoRoute>{



  String demoText='Demo app';
  bool isSep=false;
  bool isDefaultPressed=false;
  Color backgroundColor= Colors.teal;
  Color fabColor=Colors.white;
  Color fabBgColor=Colors.blue;
  double angleOffset=90;
  double visiblePart=1;
  int buttons=4;
  WheelAnimationType animationType = WheelAnimationType.around;
  WheelSize wheelSize = WheelSize.wheel_medium_120;


  @override
  void initState() {

    super.initState();
  }




  Widget faw() {
    return new FloatingActionWheel(

      buttons: [
        WheelButton(
            onPressed: () => setState(() {
                  demoText = "button #1 pressed";
                  backgroundColor = Colors.orangeAccent;
                }),
            image:  Image.asset('assets/wheel_icon.png',
              height: 35,width: 35,filterQuality:FilterQuality.high),
            backgroundColor: Colors.orange
        ),

        WheelButton(onPressed: () {
          setState(() {
            demoText = "button #2 pressed";
            backgroundColor = Colors.greenAccent;
          });
        },
            icon: Icons.ac_unit,
            backgroundColor: Colors.green),

        if(buttons > 2)
          WheelButton(onPressed: () {
            setState(() {
              demoText = "button #3 pressed";
              backgroundColor = Colors.cyanAccent;
            });
          },
              text: 'button 3',
              backgroundColor: Colors.cyan),

        if(buttons > 3)
          WheelButton(
              onPressed: () {
                setState(() {
                  demoText = "button #4 pressed";
                  backgroundColor = Colors.pinkAccent;
                });
              },
              icon: Icons.home,
              backgroundColor: Colors.pink),

        if(buttons > 4)
          WheelButton(
              onPressed: () {
                setState(() {
                  demoText = "button #5 pressed";
                  backgroundColor = Colors.purpleAccent;
                });
              },
              text: 'button 5',
              textSize: 14,
              iconSize: 29,
              icon: Icons.info,
              backgroundColor: Colors.purple),

        if(buttons > 5)
          WheelButton(
              onPressed: () {
                setState(() {
                  demoText = "button #6 pressed";
                  backgroundColor = Colors.blueAccent;
                });
              },
              // text: 'button 6',
              // icon: Icons.image,
              image: Image.asset('assets/wheel_icon.png',
                  height: 35,width: 35,filterQuality:FilterQuality.high),
              backgroundColor: Colors.blue),

      ],
      angleOffset: angleOffset,
      visiblePart: visiblePart,
      animationType: animationType,
      wheelSize: wheelSize,
      fabForegroundColor: fabColor,
      fabBackgroundColor: fabBgColor,
      separated: isSep,
      defaultPressed: isDefaultPressed,
      fabElevation: 0,
    );
  }


  Widget animationSpinner(){
    return DropdownButton<WheelAnimationType>(
      onChanged: (val){setState(() {
        animationType=val??WheelAnimationType.none;
      });},
      value: animationType,
      disabledHint:  const Text("select wheel animation"),
      items: [
        DropdownMenuItem(
          child:  const Text("Around"),
          value: WheelAnimationType.around,
        ),
        DropdownMenuItem(
          child:  const Text("Center"),
          value: WheelAnimationType.center,
        ),
        DropdownMenuItem(
          child:  const Text("none"),
          value: WheelAnimationType.none,
        ),
        // DropdownMenuItem(
        //   child:  const Text("Rotor",style: const TextStyle(fontSize: 12,color: Colors.red,fontStyle: FontStyle.italic),),
        //   value: WheelAnimationType.apacheRotor,
        // ),
      ],

    );}

  Widget sizeSpinner(){
    return DropdownButton<WheelSize>(
      onChanged: (val){
        setState(() {
        wheelSize=val??WheelSize.wheel_medium_120;
      });
        },
      value: wheelSize,
      disabledHint: Text("select wheel size"),
      items: [
        DropdownMenuItem(
          child:  const Text("Small"),
          value: WheelSize.wheel_small_90,
        ),
        DropdownMenuItem(
          child:  const Text("Medium"),
          value: WheelSize.wheel_medium_120,
        ),
        DropdownMenuItem(
          child:  const Text("Large"),
          value: WheelSize.wheel_large_150,
        ),
      ],

    );}

  Widget buttonsSpinner(){
    return Container(
      width: 50,
      child: DropdownButton<int>(
        onChanged: (val){setState(() {
          buttons=val??4;
        });},
        value: buttons,
        disabledHint: Text("number of buttons"),
        items: [
          DropdownMenuItem(
            child:  const Text("2"),
            value: 2,
          ),
          DropdownMenuItem(
            child:  const Text("3"),
            value: 3,
          ),
          DropdownMenuItem(
            child:  const Text("4"),
            value: 4,
          ),
          DropdownMenuItem(
            child:  const Text("5"),
            value: 5,
          ),
          DropdownMenuItem(
            child: const  Text("6"),
            value: 6,
          ),
        ],

      ),
    );}



  Widget demoContainer() {
    return Container(

      color: Color.fromARGB(255, 230, 230, 230),
      child: Padding(
        padding: const EdgeInsets.all(28.0),
        child: Container(
          decoration: BoxDecoration(
              borderRadius:  const BorderRadius.all(Radius.elliptical(50,90)),
              color: backgroundColor
          ),
          child: Center(
            child: Padding(
              padding: const EdgeInsets.all(15.0),
              child: Text(demoText),
            ),
          ),

        ),
      ),
    );
  }


  @override
  Widget build(BuildContext context) {


    Widget menusContainer() {
      return Container(
        child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    const Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Wheel active part:'),
                    ),
                    Container(
                      width: 50,
                      child: TextField(
                          maxLines: 1,
                          maxLength: 3,
                          // focusNode: FocusNode(canRequestFocus: true),
                          // onEditingComplete: (){
                          // },
                          autofocus: false,
                          keyboardType: TextInputType.number,
                          decoration: InputDecoration(
                              hintText: '0.1-1',
                              border: UnderlineInputBorder(),
                              counterText: ''
                          ),
                          onChanged: (text) {
                            if (double.tryParse(text) != null)
                              if (double.tryParse(text)! >= 0.1 &&
                                  double.tryParse(text)! <= 1)
                                setState(() {
                                  visiblePart = double.tryParse(text)!;
                                });

                          }),

                    ),
                     Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Angle offset:'),
                    ),
                    Container(
                      width: 50,
                      child: TextField(
                          autofocus: false,
                          // focusNode: FocusNode(canRequestFocus: true),
                          maxLines: 1,
                          maxLength: 3,
                          decoration: InputDecoration(
                              hintText: '0-360',
                              border: UnderlineInputBorder(),
                              counterText: ''
                          ),
                          keyboardType: TextInputType.number,
                          onChanged: (text) {
                            if (double.tryParse(text) != null)
                              if (double.tryParse(text)! >= 0 &&
                                  double.tryParse(text)! <= 360)
                                setState(() {
                                  angleOffset = double.tryParse(text)!;
                                });

                          }),
                    ),
                  ],
                ),
              ),
              SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    const  Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Wheel buttons:'),
                    ),
                    buttonsSpinner(),
                    const Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Separating bezels:',style: const TextStyle(fontSize: 14),),
                    ),
                    Container(
                      child:
                      /* could make isSeparated boolean in a DropDownMenu instead of Switch because
                      popup menu overlays trigger all overlays rebuild which rebuilds the ImmersiveWheel overlay immediately */
                      // DropdownButton<bool>(
                      //   onChanged: (val){setState(() {
                      //     isSep=val!;
                      //   });},
                      //   value: isSep,
                      //   disabledHint: Text("separated"),
                      //   items: [
                      //     DropdownMenuItem(
                      //       child:  const Text("false"),
                      //       value: false,
                      //     ),
                      //     DropdownMenuItem(
                      //       child:  const Text("true"),
                      //       value: true,
                      //     ),
                      //   ],
                      //
                      // ),
                        Switch(
                          value: isSep,
                          activeColor: Colors.blue,
                          onChanged: (_) =>
                            setState(() {
                              isSep = !isSep;
                            }),
                        )
                    )
                  ],),
              ),
              SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    const Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Wheel animation type:'),
                    ),
                    animationSpinner(),
                  ],
                ),
              ),
              SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    Padding(
                      padding: const EdgeInsets.all(8.0),
                      child:  const Text('Wheel size:'),
                    ),
                    sizeSpinner(),
                  ],
                ),
              ),
              const  Divider(color: Colors.grey,height: 0,thickness: 1,)
            ]),
      );
    }



    // print('demo rebuild');


    return Scaffold(
      appBar: AppBar(
        title: const  Text('Floating Action Wheel Demo App'),

      ),
     // floatingActionButton: faw(),
     // floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,
//      bottomNavigationBar: BottomAppBar(
//        shape: CircularNotchedRectangle(),
////          notchMargin: 100,
//        elevation: 20,
//        child: Row(
//          children: [
//            Icon(Icons.home,size: 50,)
//          ],
//        ),
//      ),
      body: SingleChildScrollView(
        child: Column(

          children: [
            menusContainer(),
            demoContainer(),
            Container(
              height: 320,
              color: const Color.fromARGB(255, 230, 230, 230),
              child: Center(
                child: faw(),
              ),
            )
          ],
        ),
      ),
    );


  }
}

